python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -0,0 +1,342 @@
+        # was created.
+        self.name = "{}@{}".format(type(self).__qualname__, time.time())
+        log.debug("Creating datastore %s", self.name)
+
+        # Scan for child datastores and instantiate them with the same registry
+        self.datastores = []
+        for c in self.config["datastores"]:
+            c = DatastoreConfig(c)
+            datastoreType = doImport(c["cls"])
+            datastore = datastoreType(c, registry)
+            log.debug("Creating child datastore %s", datastore.name)
+            self.datastores.append(datastore)
+
+    def exists(self, ref):
+        """Check if the dataset exists in one of the datastores.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required dataset.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the entity exists in one of the child datastores.
+        """
+        for datastore in self.datastores:
+            if datastore.exists(ref):
+                log.debug("Found %s in datastore %s", ref, datastore.name)
+                return True
+        return False
+
+    def get(self, ref, parameters=None):
+        """Load an InMemoryDataset from the store.
+
+        The dataset is returned from the first datastore that has
+        the dataset.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+        parameters : `dict`
+            `StorageClass`-specific parameters that specify, for example,
+            a slice of the Dataset to be loaded.
+
+        Returns
+        -------
+        inMemoryDataset : `object`
+            Requested Dataset or slice thereof as an InMemoryDataset.
+
+        Raises
+        ------
+        FileNotFoundError
+            Requested dataset can not be retrieved.
+        TypeError
+            Return value from formatter has unexpected type.
+        ValueError
+            Formatter failed to process the dataset.
+        """
+
+        for datastore in self.datastores:
+            try:
+                inMemoryObject = datastore.get(ref, parameters)
+                log.debug("Found Dataset %s in datastore %s", ref, datastore.name)
+                return inMemoryObject
+            except FileNotFoundError:
+                pass
+
+        raise FileNotFoundError("Dataset {} could not be found in any of the datastores".format(ref))
+
+    def put(self, inMemoryDataset, ref):
+        """Write a InMemoryDataset with a given `DatasetRef` to each
+        datastore.
+
+        Parameters
+        ----------
+        inMemoryDataset : `object`
+            The Dataset to store.
+        ref : `DatasetRef`
+            Reference to the associated Dataset.
+
+        Raises
+        ------
+        TypeError
+            Supplied object and storage class are inconsistent.
+        """
+
+        for datastore in self.datastores:
+            datastore.put(inMemoryDataset, ref)
```

Comment:
What does a `Datastore` normally do when it isn't configured to support a particular `DatasetRef` when it's asked to `put` (e.g. a `PosixDatastore` that doesn't have a formatter configured)?

If the answer is "raise an exception", we probably either want to change the behavior of `ChainedDatastore` itself to ignore those failures, or add a new trivial forwarding `Datastore` in between that does.  I think it'll be pretty common to want to write only some `DatasetTypes` to certain `Datastores` in the chain.

Reply:
I agree that if a datastore put fails in a specific way, we can skip that datastore and allow the chained put to work if at least one put worked (possibly to a non-ephemeral datastore -- we could make that configurable).